BEIMA.Backend.Test/ReportService/ReportServiceTest.cs

```diff
@@ -0,0 +1,435 @@
+        {
+            var deviceTypeOne = new DeviceType(ObjectId.GenerateNewId(), "Boiler", "This is a boiler", "Boiler type notes");
+            deviceTypeOne.SetLastModified(DateTime.UtcNow, "Anonymous");
+            deviceTypeOne.AddField("boilerf1", "BoilerField1");
+            deviceTypeOne.AddField("boilerf2", "BoilerField2");
+            deviceTypeOne.AddField("boilerf3", "BoilerField3");
+
+            var deviceOne = new Device(ObjectId.GenerateNewId(), deviceTypeOne.Id, "dTag1", "dMan1", "dMod1", "dSer1", 2001, "dNote1");
+            deviceOne.SetLastModified(DateTime.UtcNow, "Anonymous");
+            deviceOne.SetLocation(new ObjectId("111111111111111111111111"), "dLocNotes1", "1", "1");
+            deviceOne.AddField("boilerf1", "BoilerValue1D1");
+            deviceOne.AddField("boilerf2", "BoilerValue2D1");
+            deviceOne.AddField("boilerf3", "BoilerValue3D1");
+
+            var deviceTwo = new Device(ObjectId.GenerateNewId(), deviceTypeOne.Id, "dTag2", "dMan2", "dMod2", "dSer2", 2002, "dNote2");
+            deviceTwo.SetLastModified(DateTime.UtcNow, "Anonymous");
+            deviceTwo.SetLocation(new ObjectId("111111111111111111111111"), "dLocNotes2", "2", "2");
+            deviceTwo.AddField("boilerf1", "BoilerValue1D2");
+            deviceTwo.AddField("boilerf2", "BoilerValue2D2");
+            deviceTwo.AddField("boilerf3", "BoilerValue3D2");
+
+            var deviceDocs = new List<BsonDocument>()
+            {
+                deviceOne.GetBsonDocument(),
+                deviceTwo.GetBsonDocument(),
+            };
+
+            // Setup mock database client.
+            Mock<IMongoConnector> mockDb = new Mock<IMongoConnector>();
+            mockDb.Setup(mock => mock.GetDeviceType(It.Is<ObjectId>(oid => oid == deviceTypeOne.Id)))
+                  .Returns(deviceTypeOne.GetBsonDocument())
+                  .Verifiable();
+            mockDb.Setup(mock => mock.GetFilteredDevices(It.IsAny<FilterDefinition<BsonDocument>>()))
+                  .Returns(deviceDocs)
+                  .Verifiable();
+            MongoDefinition.MongoInstance = mockDb.Object;
+
+            var reportService = ReportWriter.Instance;
+
+            // ACT
+            var bytes = reportService.GeneratDeviceReportByDeviceType(deviceTypeOne.Id);
+
+            // Assert
+            Assert.That(bytes, Is.Not.Null);
+            Assert.DoesNotThrow(() => mockDb.Verify(mock => mock.GetDeviceType(It.Is<ObjectId>(oid => oid == deviceTypeOne.Id)), Times.Once));
+
+            var content = Encoding.UTF8.GetString(bytes);
+            var contentRows = content.Split(Environment.NewLine);
+            Assert.That(contentRows.Count, Is.EqualTo(3));
+
+            // Tests that first row is a header row
+            Assert.That(GenerateDeviceHeaders(deviceTypeOne), Is.EqualTo(contentRows[0]));
+            Assert.That(GenerateDeviceValues(deviceOne, deviceTypeOne), Is.EqualTo(contentRows[1]));
```

Comment:
I should note that the formatting for this is meant to be Assert.That(actual, Is.EqualTo(expected)), instead of the other way around. Otherwise the error message it produces on failures is a little confusing.